flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMasterServiceLeadershipRunner.java

```diff
@@ -277,18 +277,11 @@ private void startJobMasterServiceProcessAsync(UUID leaderSessionId) {
 
     @GuardedBy("lock")
     private void verifyJobSchedulingStatusAndCreateJobMasterServiceProcess(UUID leaderSessionId)
-            throws FlinkException {
-        try {
-            if (jobResultStore.hasJobResultEntry(getJobID())) {
-                jobAlreadyDone(leaderSessionId);
-            } else {
-                createNewJobMasterServiceProcess(leaderSessionId);
-            }
-        } catch (IOException e) {
-            throw new FlinkException(
-                    String.format(
-                            "Could not retrieve the job scheduling status for job %s.", getJobID()),
-                    e);
+            throws FlinkException, ExecutionException, InterruptedException {
+        if (jobResultStore.hasJobResultEntryAsync(getJobID()).get()) {
```

Comment:
This one, I feel like addressing in a separate review because it requires some deeper changes. But we could leverage the async nature here as well.

Reply:
Yes, I agree that make both `startJobMasterServiceProcessAsync` and `verifyJobSchedulingStatusAndCreateJobMasterServiceProcess` leverage the async nature is important. 
Currently dealing with the sync and async future logic in these methods will be complex and make the code hard to understand, so refactoring this needs to be well designed and may be completed in another pr, WDYT?